### PR TITLE
Add ShoppingCart model

### PR DESCRIPTION
**Pull request by 4Vitalii5**
Spring Boot practice
At the end of this HW you should have an ability to add a certain book to your shopping cart, so you need to implement the missing functionality. Below you will see the detailed requirements what should be in your project.

User Use Cases (means these operations are allowed to users with role USER)
Use cases are the set of actions that could be performed by some actor (in this case user). Here are the list of use cases that we will cover in this part of HW:

Shopping Cart Management:

As a User, I want to add books to my shopping cart, so I can purchase them. I will:
Send a POST request to /api/cart to add a book to the shopping cart.
As a User, I want to view my shopping cart, so I can review its contents before placing an order. I will:
Send a GET request to /api/cart to retrieve my shopping cart.
As a User, I want to update the books quantity in my shopping cart, so I can update my intended purchases. I will:
Send a PUT request to /api/cart/items/{cartItemId} to update the books quantity in the shopping cart.
As a User, I want to remove books from my shopping cart, so I can update my intended purchases. I will:
Send a DELETE request to /api/cart/items/{cartItemId} to remove a book from the shopping cart.
Domain models (entities)
There is a list of all entities that should be present in the project after this HW:

Book: Represents a book available in the store.
User: Contains information about the registered user including their authentication details and personal information.
Role: Represents the role of a user in the system, for example, admin or user.
Category: Represents a category that a book can belong to.
ShoppingCart: Represents a user's shopping cart.
CartItem: Represents an item in a user's shopping cart.
ShoppingCart entity
ShoppingCart: Represents a user's shopping cart. You can use soft delete for shopping carts.

Add a new entity ShoppingCart with the next fields:
id (Long, PK)
user (User, not null)
cartItems (Set<CartItem>)
CartItem entity
CartItem: Represents an item in a user's shopping cart. You can avoid using soft delete for cart items and remove them permanently.

Add a new entity CartItem with the next fields:
id (Long, PK)
shoppingCart (ShoppingCart, not null)
book (Book, not null)
quantity (int, not null)
HINT: It may be helpful to add @Named("bookFromId") default Book bookFromId(Long id) { // your implementation here} to the BookMapper interface

### CODE REVIEW

File: src/main/java/mate/academy/dto/item/CreateCartItemRequestDto.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.NotNull;
+
+public record CreateCartItemRequestDto(
+        @NotNull
+        @Min(0)
+        Long bookId,
+        @NotNull
+        @Min(0)
+        int quantity
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/dto/item/CreateCartItemRequestDto.java]: int type can't be null

```suggestion
        @Positive
        Long bookId,
        @Positive
        int quantity
```

File: src/main/java/mate/academy/dto/item/UpdateCartItemDto.java [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package mate.academy.dto.item;
+
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.NotNull;
+
+public record UpdateCartItemDto(
+        @NotNull
+        @Min(1)
+        int quantity
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/dto/item/UpdateCartItemDto.java]:
```suggestion
        @Positive
        int quantity
```

File: src/main/java/mate/academy/dto/user/UserLoginRequestDto.java [hunk 3]
```diff
@@ -6,11 +6,11 @@
 
 public record UserLoginRequestDto(
         @NotEmpty
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/dto/user/UserLoginRequestDto.java]:
```suggestion
        @NotBlank
```

File: src/main/java/mate/academy/dto/user/UserLoginRequestDto.java [hunk 4]
```diff
@@ -6,11 +6,11 @@
...
-        @Length(min = 8, max = 20)
+        @Length(min = 8, max = 60)
         @Email
         String email,
         @NotEmpty
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/dto/user/UserLoginRequestDto.java]:
```suggestion
        @NotBlank
```

File: src/main/java/mate/academy/service/impl/ShoppingCartServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    @Override
+    public void createShoppingCart(User user) {
+        ShoppingCart shoppingCart = new ShoppingCart();
+        shoppingCart.setUser(user);
+        shoppingCartRepository.save(shoppingCart);
+    }
+
+    @Override
+    public ShoppingCartDto getShoppingCart() {
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(getCurrentUserId());
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
        ShoppingCart shoppingCart = getShoppingCartByCurrentUser();
```

File: src/main/java/mate/academy/service/impl/ShoppingCartServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
+    public ShoppingCartDto addBookToShoppingCart(CreateCartItemRequestDto requestDto) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(getCurrentUserId());
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
        ShoppingCart shoppingCart = getShoppingCartByCurrentUser();
```

File: src/main/java/mate/academy/service/impl/ShoppingCartServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+        shoppingCartRepository.save(shoppingCart);
+    }
+
+    @Override
+    public ShoppingCartDto getShoppingCart() {
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(getCurrentUserId());
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/service/impl/ShoppingCartServiceImpl.java:46]:
```suggestion
    @Override
    @Transactional
```

File: src/main/java/mate/academy/service/impl/ShoppingCartServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+    public ShoppingCartDto addBookToShoppingCart(CreateCartItemRequestDto requestDto) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(getCurrentUserId());
+        addOrUpdateCartItem(requestDto, shoppingCart);
+        shoppingCartRepository.save(shoppingCart);
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/service/impl/ShoppingCartServiceImpl.java:55]:
```suggestion
    @Override
    @Transactional
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/mate/academy/service/impl/ShoppingCartServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+        if (existingCartItem.isPresent()) {
+            int updatedQuantity = existingCartItem.get().getQuantity() + requestDto.quantity();
+            existingCartItem.get().setQuantity(updatedQuantity);
+        } else {
+            CartItem newCartItem = cartItemMapper.toEntity(requestDto);
+            newCartItem.setBook(book);
+            newCartItem.setShoppingCart(shoppingCart);
+            shoppingCart.getCartItems().add(newCartItem);
+            cartItemRepository.save(newCartItem);
+        }
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/service/impl/ShoppingCartServiceImpl.java:105]: better avoid if-else construction  
```suggestion
        if (existingCartItem.isPresent()) {
            int updatedQuantity = existingCartItem.get().getQuantity() + requestDto.quantity();
            existingCartItem.get().setQuantity(updatedQuantity);
            return;
        } 
            CartItem newCartItem = cartItemMapper.toEntity(requestDto);
            newCartItem.setBook(book);
            newCartItem.setShoppingCart(shoppingCart);
            shoppingCart.getCartItems().add(newCartItem);
            cartItemRepository.save(newCartItem);
```

**Review by liliia-ponomarenko — APPROVED** [review]
awesome 😎

File: src/main/java/mate/academy/service/impl/ShoppingCartServiceImpl.java (hunk 9 above)
Comment (4Vitalii5) [reply src/main/java/mate/academy/service/impl/ShoppingCartServiceImpl.java:105]: Thank you.

**Pull request merged**